src/java/org/apache/cassandra/locator/AbstractReplicaCollection.java

```diff
@@ -58,45 +63,332 @@
+            {
+                final int end = begin + size;
+                int i = begin;
+                @Override
+                public boolean hasNext()
+                {
+                    return i < end;
+                }
+
+                @Override
+                public K next()
+                {
+                    return function.apply(contents[i++]);
+                }
+            };
+        }
+
+        private Iterator<Replica> filterIterator(Predicate<Replica> predicate, int limit)
+        {
+            return new Iterator<Replica>()
+            {
+                final int end = begin + size;
+                int next = begin;
+                int count = 0;
+                { updateNext(); }
+                void updateNext()
+                {
+                    if (count == limit) next = end;
+                    while (next < end && !predicate.test(contents[next]))
+                        ++next;
+                    ++count;
+                }
+                @Override
+                public boolean hasNext()
+                {
+                    return next < end;
+                }
+
+                @Override
+                public Replica next()
+                {
+                    if (!hasNext()) throw new IllegalStateException();
+                    Replica result = contents[next++];
+                    updateNext();
+                    return result;
+                }
+            };
+        }
+
+        @VisibleForTesting
+        public boolean equals(Object to)
+        {
+            if (to == null || to.getClass() != ReplicaList.class)
+                return false;
+            ReplicaList that = (ReplicaList) to;
+            if (this.size != that.size) return false;
+            for (int i = 0 ; i < size ; ++i)
+                if (!this.get(i).equals(that.get(i)))
+                    return false;
+            return true;
+        }
     }
 
-    // if subList == null, should return self (or a clone thereof)
-    protected abstract C snapshot(List<Replica> subList);
-    protected abstract C self();
     /**
-     * construct a new Mutable of our own type, so that we can concatenate
-     * TODO: this isn't terribly pretty, but we need sometimes to select / merge two Endpoints of unknown type;
+     * A simple map that ensures the underlying list's iteration order is maintained, and can be shared with
+     * subLists (either produced via subList, or via filter that naturally produced a subList).
+     * This permits us to reduce the amount of garbage generated, by not unnecessarily copying,
+     * reduces the amount of indirection necessary, as well as ensuring monomorphic callsites.
+     * The underlying map is also more efficient, particularly for such small collections as we typically produce.
      */
-    public abstract Mutable<C> newMutable(int initialCapacity);
+    protected static class ReplicaMap<K> extends AbstractMap<K, Replica>
```

Comment:
TBH I'm not 100% sure about that idea, but if we do not inherit from `AbstractMap` (and make this class public), we can:
  * remove entryset
  * remove `remove`
  * make `get` and `containsKey` generic

Not that it's a huge difference, but it does reduce the code. Maybe there are differences from having `Map` interface, but it seems that it's a fairly internal thing and it seems that outside collections mostly `get` is used anyways.

However, we might add a comparable amount of code just in `equals` and `hashCode`. Anyways, maybe it's not a very good idea.

Reply:
Hmm.  Not sure about this.  It seems less surprising to users of the `ReplicaCollection` API if they get a regular plain `Map`, and it's a nominal amount of code.

We also don't strictly *need* `remove` - I added it only to ensure consistent `UnsupportedOperationException` - otherwise we only throw when we try to remove a *present* value, as opposed to for all invocations of `remove`.

It's a good point, though, as we could simply not expose the `Map` at all, and replace `byEndpoint()` with simply `lookupByEndpoint()` - keeping the `KeySet` for implementing our `endpoints()` methods (and similarly for the range equivalents).  

I guess I'm slightly predisposed to provide the extra functionality for future users, since it's fairly low cost code wise, but it's at best a modest inclination.

Reply:
Let's leave it as is then. Thanks for considering!